mungegithub/mungers/approvers/owners.go

```diff
@@ -0,0 +1,209 @@
+	}
+	sort.Strings(approversOnly)
+	return approversOnly
+}
+
+
+// Shuffles the potential approvers so that we don't always suggest the same people
+func (o Owners) GetShuffledApprovers() []string {
+	approversList := o.GetAllPotentialApprovers()
+	order := rand.Perm(len(approversList))
+	people := make([]string, 0, len(approversList))
+	for i := range order{
+		people = append(people, approversList[i])
+	}
+	return people
+}
+
+
+// findPersonWithBiggestCoverage finds the person who is capable of approving
+// the greatest number of relevant OWNERS files
+func findPersonWithBiggestCoverage(toBeApproved sets.String, randomizedApprovers []string, reverseMap map[string]sets.String) string {
+	maxCovered := 0
+	var bestPerson string
+	for _, approver := range randomizedApprovers {
+		filesCanApprove := reverseMap[approver]
+		if filesCanApprove.Intersection(toBeApproved).Len() > maxCovered {
+			maxCovered = len(filesCanApprove)
+			bestPerson = approver
+		}
+	}
+	return bestPerson
+
+}
+// GetSuggestedApprovers solves the exact cover problem, finding an approver capable of
+// approving every OWNERS file in the PR
+func (o Owners) GetSuggestedApprovers() []string {
+
+	randomizedApprovers := o.GetShuffledApprovers()
+	revMap := o.GetReverseMap()
+
+	setToCover := sets.NewString()
+	for _, fn := range o.filenames {
+		setToCover.Insert(fn)
+	}
+
+	approverGroup := sets.NewString()
+	for setToCover.Len() > 0 {
+		bestPerson := findPersonWithBiggestCoverage(setToCover, randomizedApprovers, revMap)
+		approverGroup.Insert(bestPerson)
+		toDelete := sets.NewString()
+		// remove all files in the directories that our approver approved AND
+		// in the subdirectories that s/he approved.  HasPrefix finds subdirs
+		for fn := range setToCover {
+			for approvedFile := range revMap[bestPerson]{
+				if strings.HasPrefix(fn, approvedFile) {
+					toDelete.Insert(fn)
+				}
+			}
+		}
+		setToCover.Delete(toDelete.List()...)
+	}
+	toSuggest := approverGroup.List()
+	return toSuggest
+
+}
+
+// removeSubdirs takes a list of directories as an input and returns a set of directories with all
+// subdirectories removed.  E.g. [/a,/a/b/c,/d/e,/d/e/f] -> [/a, /d/e]
+func removeSubdirs(dirList []string) sets.String {
+	toDel := sets.String{}
+	for i := 0; i < len(dirList)-1; i++ {
+		for j := i + 1; j < len(dirList); j++ {
+			// ex /a/b has prefix /a so if remove /a/b since its already covered
+			if strings.HasPrefix(dirList[i], dirList[j]) {
+				toDel.Insert(dirList[i])
+			} else if strings.HasPrefix(dirList[j], dirList[i]) {
+				toDel.Insert(dirList[j])
+			}
+		}
+	}
+	finalSet := sets.NewString(dirList...)
+	finalSet.Delete(toDel.List()...)
+	return finalSet
+}
+
+// GetOwnersSet returns a set containing all the Owners files necessary to get the PR approved
+func (o Owners) GetOwnersSet() sets.String {
+	owners := sets.NewString()
+	for _, fn := range o.filenames {
+		owners.Insert(o.repo.FindOwnersForPath(fn))
+	}
+	return removeSubdirs(owners.List())
+}
```

Comment:
I call removeSubdirs here so that we only return the relevant OWNERS files

E.g if we have /a/OWNERS and /a/b/OWNERS as the list.  We only need an approver from /a/OWNERS